specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+
+*capabilities* - Non empty list of features supported by a plugin. Each capability such as `SIGNATURE_ENVELOPE_GENERATOR` requires one of more commands to be implemented by the plugin. When new features are available for plugins to implement, an implementation may choose to not implement it, and therefore will not include the feature in capabililies. Notation will evaluate the capability required to satisfy a user’s request, and reject the request if the plugin does not support the required capability.
+
+## Signing interfaces
+
+### Requirements
+
+* The interface MUST NOT be tied to a specific signature envelope format.
+
+Notation will support plugins to be developed against the following interfaces - *Signature Generator*, and *Signature Envelope Generator*. These interfaces target abstraction levels that satisfy most plugin integration scenarios.
+
+### Signature Generator
+
+This interface targets plugins that integrate with providers of basic cryptographic operations. E.g. Local PIV/PKCS#11 hardware tokens, remote KMS, or key vault services. Plugins that target this interface will only generate a raw signature given a payload to sign. Notation will package this signature into a signature envelope, and generate the signature manifest. Notation will also generate the TSA signature if indicated by the user. The plugin does not need to be signature envelope format aware, and will continue to work if Notary adopts additional signature formats.
+
+#### Signing workflow using plugin
+
+1. Given a user request to sign `oci-artifact`, with `key`
+1. Pull the image manifest using `oci-artifact` url, and construct a descriptor
+1. Append any user provided metadata and Notary metadata as descriptor annotations.
+1. Determine if the registered key uses a plugin
+1. Execute the plugin with `discover` command
+    1. If plugin supports capability `SIGNATURE_GENERATOR`
+        1. Generate the payload to be signed. For JWS this includes [JWS](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#supported-signature-envelopes) payload with `subject` claim as descriptor, additional JWS claims, and protected headers.
+        2. Execute the plugin with `generate-signature` command, set `request.key` to key definition, `request.payload` to the generated payload.
```

Comment:
`request.key` must be optional since signing keys may be private to the remote API.
You are really asking for two modes here:
1. I have the key but can't do crypto (at least not native to notation), so let plugin handle that
2. I don't the key and of course can't do crypto, so let plugin ship the payload and get it signed.

Reply:
The `request.key` is the key definition in notation config, here `name` is a friendly name to refer from CLI, `id` is the plugin/crypto specific identifier that allows plugin to know which key to use. They do not contain the value of signing key itself. I should probably change `request.key` to `request.keyDefinition` or similar to make this clearer.

``` json
"keys": [
      {  
        "name": "mysigningkey",  
        "id" : "keyid",
        "plugin": "com.example.nv2plugin"
      }  
```